File: npm-packages/meteor-capacitor/ios/Sources/MeteorjsCapacitor/CapacitorMeteorWebAppPlugin.swift
// Vendored from https://github.com/banjerluke/capacitor-meteor-webapp
// Copyright 2025 Luke Abbott. MIT license. See LICENSES/banjerluke-capacitor-meteor-webapp.txt.

//
// CapacitorMeteorWebAppPlugin.swift
//
// Main plugin class that bridges the Meteor webapp functionality
// to Capacitor, providing hot code push capabilities for iOS.
//

import Capacitor
import Foundation
import WebKit

// Bridge adapter to implement our protocol with Capacitor
class CapacitorBridgeAdapter: CapacitorBridge {
    weak var bridge: CAPBridgeProtocol?

    init(bridge: CAPBridgeProtocol?) {
        self.bridge = bridge
    }

    func setServerBasePath(_ path: String) {
        bridge?.setServerBasePath(path)
    }

    func getWebView() -> AnyObject? {
        return bridge?.webView
    }

    var webView: WKWebView? {
        return bridge?.webView
    }

    func reload() {
        // CAPBridgeProtocol doesn't have a reload method, so we reload the webView directly
        bridge?.webView?.reload()
    }
}

/// Capacitor MeteorWebApp Plugin
/// Enables hot code push functionality for Meteor apps
@objc(CapacitorMeteorWebAppPlugin)
public class CapacitorMeteorWebAppPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorMeteorWebAppPlugin"
    public let jsName = "CapacitorMeteorWebApp"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "checkForUpdates", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startupDidComplete", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCurrentVersion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isUpdateAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reload", returnType: CAPPluginReturnPromise)
    ]
    private var implementation: CapacitorMeteorWebApp!
    private var bridgeAdapter: CapacitorBridgeAdapter!

    override public func load() {
        bridgeAdapter = CapacitorBridgeAdapter(bridge: self.bridge)
        implementation = CapacitorMeteorWebApp(capacitorBridge: bridgeAdapter)

        // Listen for update notifications from the implementation
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleUpdateAvailable(_:)),
            name: .meteorWebappUpdateAvailable,
            object: nil
        )

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleUpdateFailed(_:)),
            name: .meteorWebappUpdateFailed,
            object: nil
        )
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    @objc private func handleUpdateAvailable(_ notification: Notification) {
        guard let userInfo = notification.userInfo,
              let version = userInfo["version"] as? String
        else {
            NSLog("❌ MeteorjsCapacitor: No version in notification userInfo")
            return
        }

        NSLog("🔔 MeteorjsCapacitor: Notifying JS listeners about version: \(version)")
        notifyListeners("updateAvailable", data: ["version": version])
    }

    @objc private func handleUpdateFailed(_ notification: Notification) {
        guard let userInfo = notification.userInfo,
              let errorMessage = userInfo["error"] as? String
        else { return }

        NSLog(
            "🔔 MeteorjsCapacitor: Notifying JS listeners about error: \(errorMessage)"
        )
        notifyListeners("error", data: ["message": errorMessage])
    }

    @objc func checkForUpdates(_ call: CAPPluginCall) {
        implementation.checkForUpdates { error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func startupDidComplete(_ call: CAPPluginCall) {
        implementation.startupDidComplete { error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func getCurrentVersion(_ call: CAPPluginCall) {
        let version = implementation.getCurrentVersion()
        call.resolve(["version": version])
    }

    @objc func isUpdateAvailable(_ call: CAPPluginCall) {
        let available = implementation.isUpdateAvailable()
        call.resolve(["available": available])
    }

    @objc func reload(_ call: CAPPluginCall) {
        implementation.reload { error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }
}

File: npm-packages/meteor-capacitor/android/src/main/java/com/meteorjs/capacitor/CapacitorMeteorWebAppPlugin.java
/*
 * Vendored from https://github.com/banjerluke/capacitor-meteor-webapp
 * Copyright 2025 Luke Abbott. MIT license. See LICENSES/banjerluke-capacitor-meteor-webapp.txt.
 */

package com.meteorjs.capacitor;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CapacitorMeteorWebApp")
public class CapacitorMeteorWebAppPlugin extends Plugin {

    private CapacitorMeteorWebApp implementation;
    private Throwable initializationError;

    @Override
    public void load() {
        super.load();
        try {
            implementation = new CapacitorMeteorWebApp(getBridge());
            implementation.setEventCallback(new CapacitorMeteorWebApp.EventCallback() {
                @Override
                public void onUpdateAvailable(String version) {
                    JSObject payload = new JSObject();
                    payload.put("version", version);
                    notifyListeners("updateAvailable", payload);
                }

                @Override
                public void onError(String message) {
                    JSObject payload = new JSObject();
                    payload.put("message", message);
                    notifyListeners("error", payload);
                }
            });
        } catch (Throwable cause) {
            initializationError = cause;
        }
    }

    @Override
    protected void handleOnPause() {
        super.handleOnPause();
        if (implementation != null) {
            implementation.handleOnPause();
        }
    }

    @Override
    protected void handleOnResume() {
        super.handleOnResume();
        if (implementation != null) {
            implementation.handleOnResume();
        }
    }

    @Override
    protected void handleOnDestroy() {
        super.handleOnDestroy();
        if (implementation != null) {
            implementation.handleOnDestroy();
        }
    }

    @PluginMethod
    public void checkForUpdates(PluginCall call) {
        if (!ensureReady(call)) {
            return;
        }

        implementation.checkForUpdates(error -> {
            if (error != null) {
                rejectWithThrowable(call, error);
            } else {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void startupDidComplete(PluginCall call) {
        if (!ensureReady(call)) {
            return;
        }

        implementation.startupDidComplete(error -> {
            if (error != null) {
                rejectWithThrowable(call, error);
            } else {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void getCurrentVersion(PluginCall call) {
        if (!ensureReady(call)) {
            return;
        }

        JSObject payload = new JSObject();
        payload.put("version", implementation.getCurrentVersion());
        call.resolve(payload);
    }

    @PluginMethod
    public void isUpdateAvailable(PluginCall call) {
        if (!ensureReady(call)) {
            return;
        }

        JSObject payload = new JSObject();
        payload.put("available", implementation.isUpdateAvailable());
        call.resolve(payload);
    }

    @PluginMethod
    public void reload(PluginCall call) {
        if (!ensureReady(call)) {
            return;
        }

        implementation.reload(error -> {
            if (error != null) {
                rejectWithThrowable(call, error);
            } else {
                call.resolve();
            }
        });
    }

    private boolean ensureReady(PluginCall call) {
        if (implementation != null) {
            return true;
        }

        if (initializationError != null) {
            rejectWithThrowable(
                call,
                new Exception(
                    "CapacitorMeteorWebApp failed to initialize: " + initializationError.getMessage(),
                    initializationError
                )
            );
            return false;
        }

        call.reject("CapacitorMeteorWebApp is not initialized");
        return false;
    }

    private void rejectWithThrowable(PluginCall call, Throwable error) {
        String message = error.getMessage() == null ? "Unknown error" : error.getMessage();
        if (error instanceof Exception) {
            call.reject(message, (Exception) error);
        } else {
            call.reject(message);
        }
    }
}
